src/stream.c

```diff
@@ -1391,35 +1401,44 @@ static int process_store_rules(struct stream *s, struct channel *rep, int an_bit
 	for (i = 0; i < s->store_count; i++) {
 		struct stksess *ts;
 		void *ptr;
+		char *key;
 		struct dict_entry *de;
+		struct stktable *t = s->store[i].table;
 
 		if (objt_server(s->target) && objt_server(s->target)->flags & SRV_F_NON_STICK) {
 			stksess_free(s->store[i].table, s->store[i].ts);
 			s->store[i].ts = NULL;
 			continue;
 		}
 
-		ts = stktable_set_entry(s->store[i].table, s->store[i].ts);
+		ts = stktable_set_entry(t, s->store[i].ts);
 		if (ts != s->store[i].ts) {
 			/* the entry already existed, we can free ours */
-			stksess_free(s->store[i].table, s->store[i].ts);
+			stksess_free(t, s->store[i].ts);
 		}
 		s->store[i].ts = NULL;
 
 		HA_RWLOCK_WRLOCK(STK_SESS_LOCK, &ts->lock);
-		ptr = __stktable_data_ptr(s->store[i].table, ts, STKTABLE_DT_SERVER_ID);
+		ptr = __stktable_data_ptr(t, ts, STKTABLE_DT_SERVER_ID);
 		stktable_data_cast(ptr, server_id) = __objt_server(s->target)->puid;
 		HA_RWLOCK_WRUNLOCK(STK_SESS_LOCK, &ts->lock);
 
+		if (t->server_key_type == STKTABLE_SRV_NAME)
+			key = __objt_server(s->target)->id;
+		else if (t->server_key_type == STKTABLE_SRV_ADDR)
+			key = __objt_server(s->target)->addr_desc;
+		else
+			continue;
+
 		HA_RWLOCK_WRLOCK(STK_SESS_LOCK, &ts->lock);
-		de = dict_insert(&server_name_dict, __objt_server(s->target)->id);
+		de = dict_insert(&server_key_dict, key);
```

Comment:
I'm a little bit concerned about this. I couldn't find anything that cleans up these keys, although there is a refcount on them. In an environment that uses service discovery and the servers change frequently, this could lead to a memory leak as this dict only grows in size.